### PR TITLE
[quant] AT_DISPATCH_FLOATING_AND_QINT_TYPES

### PR DESCRIPTION
**Pull request by z-a-f**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48051 [quant][refactor] Reflection pad now uses AT_DISPATCH_FLOATING_AND_QINT_TYPES
* **#48050 [quant] AT_DISPATCH_FLOATING_AND_QINT_TYPES**
* #48037 [quant] out-variant for the reflection pad
* #48036 [quant] ReflectionPad2d

The switch does not require testing -- if need to confirm if it works right, you can run the tests for the reflection padding: `python test/test_quantization.py TestPadding`

Differential Revision: [D25004159](https://our.internmc.facebook.com/intern/diff/D25004159)

### CODE REVIEW

**Review by vkuzo — APPROVED** [review]
lg, can we add a test plan to the diff summary

**Comment by z-a-f** [comment]
> lg, can we add a test plan to the diff summary

How would you test the switch?

File: aten/src/ATen/Dispatch.h [hunk 1]
```diff
@@ -280,6 +280,25 @@ inline void deprecated_AT_DISPATCH_ALL_TYPES_AND_HALF_AND_COMPLEX() {}
     }                                                                       \
   }()
 
+#define AT_DISPATCH_FLOATING_AND_QINT_TYPES(TYPE, NAME, ...)          \
```
Comment (jerryzh168) [inline aten/src/ATen/Dispatch.h:283]: oh I thought we can compose...would be even better if we can compose AT_DISPATCH_FLOATING_TYPES and AT_DISPATCH_QINT_TYPES somehow.

Comment (jerryzh168) [reply aten/src/ATen/Dispatch.h:283]: e.g. moving float cases and int cases to a macro

File: aten/src/ATen/Dispatch.h [hunk 2]
```diff
@@ -280,6 +280,25 @@ inline void deprecated_AT_DISPATCH_ALL_TYPES_AND_HALF_AND_COMPLEX() {}
...
+    at::ScalarType _st = ::detail::scalar_type(the_type);                   \
+    switch (_st) {                                                          \
+      AT_PRIVATE_CASE_TYPE(at::ScalarType::Double, double, __VA_ARGS__)     \
+      AT_PRIVATE_CASE_TYPE(at::ScalarType::Float, float, __VA_ARGS__)       \
+      AT_QINT_PRIVATE_CASE_TYPE(                                            \
+          at::kQInt8, at::qint8, at::kChar, int8_t, __VA_ARGS__)            \
+      AT_QINT_PRIVATE_CASE_TYPE(                                            \
+          at::kQUInt8, at::quint8, at::kByte, uint8_t, __VA_ARGS__)         \
+      AT_QINT_PRIVATE_CASE_TYPE(                                            \
+          at::kQInt32, at::qint32, at::kInt, int, __VA_ARGS__)              \
```
Comment (jerryzh168) [inline aten/src/ATen/Dispatch.h:296]: e.g. we might be able to compose if we move these cases to a macro, and also the float cases to another macro

File: aten/src/ATen/Dispatch.h (hunk 1 above)
Comment (z-a-f) [reply aten/src/ATen/Dispatch.h:283]: Composition would require rewriting the dispatch switches to avoid the the default case.

Comment (jerryzh168) [reply aten/src/ATen/Dispatch.h:283]: can we just define a macro for 3 AT_QINT_PRIVATE_CASE_TYPE and 2 AT_PRIVATE_CASE_TYPE ?
e.g.
```
#define QINT_CASES(__VA_ARGS__) AT_QINT_PRIVATE_CASE_TYPE(...) \
   AT_QINT_PRIVATE_CASE_TYPE(...) \
   AT_QINT_PRIVATE_CASE_TYPE(...)
```

**Pull request closed without merging**